host: Replace libvirt-lxc backend with libcontainer

lmars opened the pull request:
Opening as an early PR to see what happens in CI.

Closes #38 
Closes #3021
Closes #2553
Closes #2367
Closes #2725 
Closes #2803

lmars commented:
There are still a couple of test failures, but I think these are CI flakes present on master, so this is now ready for review.

An overview of how containers work with libcontainer:

* `flynn-host` exec's `flynn-init` with `libcontainer-init` as a command line argument, and the `_LIBCONTAINER_INITPIPE` env var pointing at a pipe file descriptor opened by `flynn-host`
* `flynn-init` receives binary netlink data followed by a JSON configuration payload from the pipe which it uses to create namespaces, setup networking etc. (this is handled transparently by `libcontainer`)
* `flynn-init` exec's `/.containerinit` inside the container (which is just a bind mount of `flynn-init` from the host), and then `containerinit` proceeds as it does currently

Some notes:

* we continue to use cgroup partitions as we did with libvirt, but these are now named `/flynn/{system,user,background}` (rather than libvirt's `/machine/{system,user,background}.partition`)
* starting a container no longer leaves a supervisor process in the root mount namespace, so we no longer need `nsumount` and I have removed it
* I have manually verified that mounts are no longer being leaked (so #38 is fixed)
* networking setup is now handled by libcontainer and has been removed from `containerinit`
* since containers are now started as direct children of `flynn-host`, we could stop passing FDs over unix sockets, but that would break restarting `flynn-host` or updating it (the containers will get re-parented to init, and we will lose the FD references).

/cc @titanous @josephglanville

titanous commented on host/flynn-init/main.go:
```diff
@@ -1,6 +1,25 @@
+		if err := factory.StartInitialization(); err != nil {
+			log.Fatal(err)
+		}
+	}
```
Is there a reason to put this in `init` instead of `main`?

titanous commented on host/libcontainer_backend.go:
```diff
@@ -455,21 +401,110 @@ func (l *LibvirtLXCBackend) Run(job *host.Job, runConfig *RunConfig, rateLimitBu
+	config := &configs.Config{
+		Rootfs: rootPath,
+		Capabilities: []string{
+			"CAP_CHOWN",
```
Where was this capabilities list taken from?

lmars replied:
This is actually the full list of capabilities from `/usr/include/linux/capability.h` on Ubuntu trusty.

I did intend to strip it down but I forgot. We could stick with the same list we have been using with `libvirt-lxc` (which blacklisted `CAP_SYS_MODULE`, `CAP_SYS_TIME`, `CAP_MKNOD`, `CAP_AUDIT_CONTROL`, and `CAP_MAC_ADMIN`, see [here](https://github.com/libvirt/libvirt/blob/v1.2.2/src/lxc/lxc_container.c#L1741-L1745)), or do a review of each capability to determine if we want to drop it or not (`libcontainer` provides "good defaults" too, see [here](https://github.com/opencontainers/runc/blob/master/libcontainer/SPEC.md#security))

titanous replied:
Let's take the list from Docker.

titanous commented on host/libcontainer_backend.go:
```diff
@@ -455,21 +401,110 @@ func (l *LibvirtLXCBackend) Run(job *host.Job, runConfig *RunConfig, rateLimitBu
+			Path: filepath.Join("/flynn", job.Partition, job.ID),
+			Resources: &configs.Resources{
+				AllowedDevices: configs.DefaultAllowedDevices,
+				Memory:         1 * units.GiB,
```
Is this the memory limit? If so, should it come from the job resource config?

titanous replied:
Nevermind, found the setting below.

titanous commented on host/libcontainer_backend.go:
```diff
@@ -602,85 +629,60 @@ func (l *LibvirtLXCBackend) Run(job *host.Job, runConfig *RunConfig, rateLimitBu
+				Address:           initConfig.IP,
+				Gateway:           initConfig.Gateway,
+				Mtu:               1500,
+				HostInterfaceName: "veth" + random.String(4),
```
This name is not random enough, it will encounter collisions (birthday paradox). What is libcontainer's behavior in this case? Perhaps we should use an atomically incrementing integer?

lmars replied:
Ah yes.

So after digging around, there is no code in libcontainer / runc which sets `HostInterfaceName` (which is why I just went with this random name). Looking at Docker v.1.11.1, it actually uses [libnetwork's GenerateIfaceName](https://github.com/docker/docker/blob/v1.11.1/vendor/src/github.com/docker/libnetwork/drivers/bridge/bridge.go#L889-L893) to do this, which just [tries a few times](https://github.com/docker/docker/blob/v1.11.1/vendor/src/github.com/docker/libnetwork/netutils/utils_linux.go#L33-L50).

We already have libnetwork vendored, so we could use that same method, or we could use an incremental int (although we will need to sychronise that across an udpate / restart). WDYT?

titanous replied:
The libnetwork function sounds fine.

titanous commented:
This is awesome! Have you tried removing the vendored libvirt client code?

titanous commented:
Looks like we'll probably need some minor docs updates.

titanous commented on host/Tupfile:
```diff
@@ -2,6 +2,5 @@ include_rules
 : cli/root_keys.go |> !cgo |> bin/flynn-host
 : bin/flynn-host |> gzip -9 --keep bin/flynn-host |> bin/flynn-host.gz
-: |> !go ./flynn-init |> bin/flynn-init
+: |> !cgo ./flynn-init |> bin/flynn-init
```
This will break containerinit if the rootfs doesn't have libc. Perhaps we should put the libcontainer part in `flynn-host` instead?

lmars commented:
@titanous comments addressed (in terms of docs, I just updated [this line](https://github.com/flynn/flynn/commit/c1c1c6a366565a36ed76b2df3f4f54bb0ac5e4eb#diff-1ecddd30aee43db4147f32e89570bad9L46) in the architecture docs).

titanous commented on host/libcontainer_backend.go:
```diff
@@ -53,6 +53,27 @@ const (
+// container, taken from:
+// https://github.com/opencontainers/runc/blob/v1.0.0-rc1/libcontainer/SPEC.md#security
+var defaultCapabilities = []string{
+	"CAP_NET_RAW",
```
This list does not appear to match the link.

lmars replied:
In what way do they differ? I just took the ones with `Enabled: 1` set.

titanous replied:
Ah, I missed the second column. That makes sense.

titanous commented on host/libcontainer_backend.go:
```diff
@@ -3,58 +3,100 @@ package main
+	// current binary with libcontainer-init as the first argument,
+	// which triggers the following code to initialise the container
+	// environment (namespaces, network etc.) then exec containerinit
+	if len(os.Args) > 1 && os.Args[1] == "libcontainer-init" {
```
Put this in the `main` function so it's obvious what's going on.

lmars commented:
Updates are not currently working because all containers are exiting when the old `flynn-host` process dies.

I initially thought `flynn-host` was killing them on shutdown but this also happens when it dies abruptly with `SIGKILL`, so it must be the kernel which kills the containers.

The likely suspect is the [ParentDeathSignal](https://github.com/opencontainers/runc/blob/v1.0.0-rc1/libcontainer/configs/config.go#L84-L86) parameter, but so far I have not been able to configure things so that the containers do not exit (I have tried modifying that config variable, removing all `prctl` calls, making a `prctl` call to reset it in `containerinit`, but none of those approaches prevent the containers from dying).

I am investigating further.

lmars commented:
The containers were previously exiting when the parent `flynn-host` daemon exited because they were getting a TTY hangup signal due to the fact that `flynn-host` opened the console, which subsequently closed when it died.

I have changed things so that only TTY jobs get a console (f5d6b84) which means those jobs will not survive an update to `flynn-host`, but those jobs are usually interactive jobs (e.g. `flynn run bash`) so I don't think we need to worry about them.

I have also fixed job resurrection so that it only happens for jobs which aren't actually running (c7d703d), and although that could be changed on master I am bundling it in here as I hope we can merge this.

The last few test runs have had some failures but they are all tests which fail intermittently on master, and are mostly controller deployment related (something which I will be refactoring soon).

@titanous @josephglanville this is ready for a review.

titanous commented on host/libcontainer_backend.go:
```diff
@@ -602,85 +636,71 @@ func (l *LibvirtLXCBackend) Run(job *host.Job, runConfig *RunConfig, rateLimitBu
-					Source: lt.FSRef{Usage: "65535"}, // 64MiB
-					Target: lt.FSRef{Dir: "/dev/shm"},
-				},
+	ifaceName, err := netutils.GenerateIfaceName("veth", 4)
```
Should this be moved into the else block below?

titanous commented on host/libcontainer_backend.go:
```diff
@@ -602,85 +636,71 @@ func (l *LibvirtLXCBackend) Run(job *host.Job, runConfig *RunConfig, rateLimitBu
+		// do anything with the console, it will be destroyed
+		// by the garbage collector and the job will get a
+		// tty hangup.
+		go io.Copy(ioutil.Discard, console)
```
This is weird, can we disassociate the console from the flynn-host process somehow?

titanous commented on script/install-flynn.tmpl:
```diff
@@ -127,7 +127,7 @@ main() {
     "aufs-tools"
     "iptables"
-    "libvirt-bin"
+    "cgroup-lite"
```
Were you able to get rid of the cgroup-lite package by setting up the cgroups in flynn-host?

lmars replied:
Ah yes forgot about this, see bea0afb.

lmars commented:
@titanous so I went on a small TTY adventure today and eventually realised the issue was that not assigning a console to the container meant that when `containerinit` [opened a PTY](https://github.com/flynn/flynn/blob/v20160707.1/host/containerinit/init.go#L673) to hand to the job, it ended up having that PTY as its controlling terminal, so then exec'ing the job with `Setctty` for that same terminal would fail with `EPERM` because terminals can only be controlling for a single session.

Long story short, updating `github.com/kr/pty` to include [this change](https://github.com/kr/pty/commit/3b1f6487b7fc649d5f146df04e623bd55ba1bf7f) fixed that, and means there is now no need to create consoles in `flynn-host` (218e56e), so therefore no TTY hang ups when `flynn-host` exits.

Ready for another review :smile:.

titanous commented:
LGTM!

josephglanville commented:
LGTM. 🎉

lmars commented:
woot!

temujin9 commented:
Woot indeed. Can I get a head's up when this releases?

titanous commented:
The latest nightly has it.
